File: src/D06.cs
namespace src06
{
  using System;
  using System.Collections.Generic;
  using System.Linq;
  using System.Reactive.Linq;

  public class OrbitMap
  {
    public OrbitMap(IEnumerable<string> definition)
    {
        directOrbits = definition.Select(x => x.Split(')')).ToDictionary(x=>x[1],x =>x[0]);
    }
    Dictionary<string,string> directOrbits;

    public int Count => directOrbits.Keys.Distinct().Select(x => ChainStartingFrom(x).Count()).Sum();
    
    public IEnumerable<string> ChainStartingFrom(string obj)
      => Observable.Generate(
          obj, o => directOrbits.Keys.Contains(o), o => directOrbits[o], o => directOrbits[o]
        ).ToEnumerable().Reverse();

    public string ClosestSharedObject(string a, string b)
      => ChainStartingFrom(a).Intersect(ChainStartingFrom(b)).Last();

    public int CountHopsFromTo(string a, string b)
    {
      var cso = ClosestSharedObject(a,b);
      return CountFrom(a)+CountFrom(b);
      int CountFrom(string o) => ChainStartingFrom(o).Reverse().TakeWhile(x => x!=cso).Count();
    }
  }
}

File: srcf/D06.fs
module Src06

open System.Linq

type OrbitMap(definition: string seq) =

  let orbitMap =
    definition
    |> Seq.map (fun s -> s.Split(')'))
    |> Seq.map (fun x -> (x.[1], x.[0]))
    |> Map.ofSeq

  member this.ChainStartingFrom(obj:string) =
    Seq.unfold (fun o -> if orbitMap.ContainsKey o then Some(o,orbitMap.[o]) else None) obj
    |> Seq.map (fun o -> orbitMap.[o])

  member this.CountAllOrbits
    with get() = orbitMap
      |> Map.toSeq
      |> Seq.map (fun x -> this.ChainStartingFrom(fst x) |> Seq.length)
      |> Seq.sum

  member this.ClosestSharedObject (a:string) (b:string) =
    let chainA = this.ChainStartingFrom a
    let chainB = this.ChainStartingFrom b
    chainA.Intersect(chainB) |> Seq.head

  member this.CountHopsFromTo (a:string) (b:string) =
    let cso = this.ClosestSharedObject a b
    let CountFrom (o:string) =
      this.ChainStartingFrom o
      |> Seq.takeWhile (fun x -> x<>cso)
      |> Seq.length
    (CountFrom a)+(CountFrom b)
